pkg/handlers/lifecycle/konnectoragent/handler.go

```diff
@@ -122,6 +135,66 @@ func (n *DefaultKonnectorAgent) BeforeClusterUpgrade(
 	req *runtimehooksv1.BeforeClusterUpgradeRequest,
 	resp *runtimehooksv1.BeforeClusterUpgradeResponse,
 ) {
+	clusterKey := ctrlclient.ObjectKeyFromObject(&req.Cluster)
+	log := ctrl.LoggerFrom(ctx).WithValues("cluster", clusterKey)
+
+	// Check if konnectorAgent is enabled before performing any operations
+	varMap := variables.ClusterVariablesToVariablesMap(req.Cluster.Spec.Topology.Variables)
+	_, err := variables.Get[apivariables.NutanixKonnectorAgent](
+		varMap,
+		n.variableName,
+		n.variablePath...)
+	if err != nil {
+		if variables.IsNotFoundError(err) {
+			log.Info("Konnector Agent addon not enabled, skipping all upgrade operations")
+			resp.SetStatus(runtimehooksv1.ResponseStatusSuccess)
+			return
+		}
+		log.Error(err, "Failed to read Konnector Agent variable")
+		resp.SetStatus(runtimehooksv1.ResponseStatusSuccess)
+		resp.SetMessage(fmt.Sprintf("failed to read Konnector Agent variable: %v", err))
+		return
+	}
+
+	// Check if legacy helm release deletion is already in progress
+	cleanupStatus, statusMsg, err := n.checkLegacyHelmReleaseDeletionStatus(ctx, &req.Cluster, log)
+	if err != nil {
+		log.Error(err, "Failed to check legacy helm release deletion status")
+		// Continue with deletion attempt on error
+		cleanupStatus = cleanupStatusNotStarted // Treat error as not started
+	}
+
+	switch cleanupStatus {
+	case cleanupStatusCompleted:
+		log.Info("Legacy helm release deletion already completed")
+		// Skip deletion since it's already completed
+	case cleanupStatusInProgress:
+		log.Info("Legacy helm release deletion in progress, requesting retry", "details", statusMsg)
+		resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+		resp.SetRetryAfterSeconds(10) // Retry after 10 seconds
+		resp.SetMessage(fmt.Sprintf(
+			"Legacy helm release deletion in progress. Waiting for deletion to complete. %s",
+			statusMsg,
+		))
+		return
+	case cleanupStatusTimedOut:
+		log.Error(
+			fmt.Errorf("legacy helm release deletion timed out"),
+			"Legacy helm release deletion timed out, continuing with upgrade",
+			"details",
+			statusMsg,
+		)
+		// Continue with upgrade even if deletion timed out - this is a best-effort cleanup
+	case cleanupStatusNotStarted:
+		log.Info("Starting legacy helm release deletion")
+		// Delete legacy helm release "nutanix-k8s-agent" if it exists.
+		// This is a best-effort cleanup operation - errors are logged but don't block the upgrade.
```

Comment:
Is this really valid, can it be best effort? From the PR description you mentioned conflicts and other reasons why we can't have both charts installed.

I don't think we want to be in this indeterministic state where down the line after a few upgrades the addon stops working because of some new conflict.